### ISSUE TITLE
feat(search): Add LLM-based query expansion for improved recall

### ISSUE THREAD

**Issue by windoliver** [nexi-lab/nexus#1174]
## Summary
Implement LLM-based query expansion to improve search recall by generating multiple query variants that feed into both BM25 and vector search pipelines.

## Background
[QMD (tobi/qmd)](https://github.com/tobi/qmd) demonstrates that query expansion with a fine-tuned 1.7B model achieves **91.7% evaluation score** on search quality. Their approach generates three types of expansions:

1. **Lexical variants (`lex:`)** - Keyword variations optimized for BM25
2. **Vector variants (`vec:`)** - Natural language reformulations for embeddings
3. **HyDE passages (`hyde:`)** - Hypothetical document passages (answer-like text)

## Architecture

```
Original Query
     │
     ▼
┌──────────────────────────────┐
│   Query Expansion (LLM)      │
│   ───────────────────────    │
│   lex: kubernetes pod restart│
│   lex: k8s deployment debug  │
│   vec: How to fix Kubernetes │
│   hyde: When a K8s deploy... │
└──────────────────────────────┘
     │
     ├── lex: variants ──► BM25 Search (multiple queries)
     │
     ├── vec: variants ──► Vector Search (multiple embeddings)
     │
     └── hyde: passages ──► Vector Search (document-like embeddings)
     │
     ▼
┌──────────────────────────────┐
│   RRF Fusion (existing)      │
│   + Position-aware bonuses   │
└──────────────────────────────┘
     │
     ▼
   Results
```

### Phase 1: Basic Query Expansion
Create `src/nexus/search/query_expansion.py`:

```python
from dataclasses import dataclass
from enum import Enum

class ExpansionType(Enum):
    LEX = "lex"   # Lexical variants for BM25
    VEC = "vec"   # Natural language for embeddings
    HYDE = "hyde" # Hypothetical document passages

@dataclass
class QueryExpansion:
    expansion_type: ExpansionType
    text: str
    weight: float = 1.0

class QueryExpander:
    """LLM-based query expansion for improved recall."""
    
    def __init__(
        self,
        model: str = "gpt-4o-mini",  # Fast & cheap
        max_expansions: int = 6,     # 2 lex + 2 vec + 2 hyde
    ):
        self.model = model
        self.max_expansions = max_expansions
    
    async def expand(
        self,
        query: str,
        context: str | None = None,
    ) -> list[QueryExpansion]:
        """Generate query expansions using LLM."""
        prompt = self._build_prompt(query, context)
        response = await self._call_llm(prompt)
        return self._parse_expansions(response)
    
    def _build_prompt(self, query: str, context: str | None) -> str:
        return f"""Generate search query expansions for: "{query}"

Output exactly 6 lines in this format:
lex: <keyword variant 1>
lex: <keyword variant 2>
vec: <natural language reformulation 1>
vec: <natural language reformulation 2>
hyde: <hypothetical document passage 1>
hyde: <hypothetical document passage 2>

Rules:
- lex: Short keyword phrases (2-5 words), include synonyms
- vec: Full sentences, natural questions
- hyde: 1-2 sentences describing what a relevant document would contain
- Preserve important entities (names, acronyms, technical terms)
- No explanations, just the expansions
{f'Context: {context}' if context else ''}"""
```

### Phase 2: Smart Triggering
Only expand when initial BM25 signal is weak (like QMD):

```python
async def search_with_expansion(
    self,
    query: str,
    limit: int = 10,
) -> list[SearchResult]:
    # Initial BM25 search
    initial = await self.bm25_search(query, limit=5)
    
    # Check signal strength
    if self._has_strong_signal(initial):
        # Skip expansion - good results already
        return await self.hybrid_search(query, limit)
    
    # Weak signal - expand query
    expansions = await self.expander.expand(query)
    
    # Run expanded searches
    lex_results = await asyncio.gather(*[
        self.bm25_search(e.text, limit=limit)
        for e in expansions if e.expansion_type == ExpansionType.LEX
    ])
    
    vec_results = await asyncio.gather(*[
        self.vector_search(e.text, limit=limit)
        for e in expansions if e.expansion_type in (ExpansionType.VEC, ExpansionType.HYDE)
    ])
    
    # RRF fusion with position-aware bonuses
    return self.rrf_fusion(
        [initial] + lex_results + vec_results,
        weights=[2.0] + [1.0] * len(lex_results) + [1.0] * len(vec_results)
    )

def _has_strong_signal(self, results: list[SearchResult]) -> bool:
    """Check if top result is strong (score >= 0.85 with separation)."""
    if not results:
        return False
    top_score = results[0].score
    second_score = results[1].score if len(results) > 1 else 0
    return top_score >= 0.85 and (top_score - second_score) >= 0.1
```

### Phase 3: Caching
Cache expansions to avoid redundant LLM calls:

```python
class CachedQueryExpander(QueryExpander):
    """Query expander with content-hash caching."""
    
    def __init__(self, cache: Redis | Dragonfly, ttl: int = 3600):
        super().__init__()
        self.cache = cache
        self.ttl = ttl
    
    async def expand(self, query: str, context: str | None = None) -> list[QueryExpansion]:
        cache_key = f"qexp:{hashlib.sha256(f'{query}:{context}'.encode()).hexdigest()[:16]}"
        
        cached = await self.cache.get(cache_key)
        if cached:
            return self._deserialize(cached)
        
        expansions = await super().expand(query, context)
        await self.cache.setex(cache_key, self.ttl, self._serialize(expansions))
        return expansions
```

### 1. Hybrid Search (`fusion.py`)
Add expansion before fusion:

```python
class HybridSearch:
    def __init__(self, ..., expander: QueryExpander | None = None):
        self.expander = expander
    
    async def search(self, query: str, ...) -> list[SearchResult]:
        if self.expander:
            return await self._search_with_expansion(query, ...)
        return await self._search_direct(query, ...)
```

### 2. Query Router (`query_router.py`)
Use expansion for complex queries:

```python
# In QueryRouter.route()
if complexity >= 0.6 and self.expander:
    strategy.use_expansion = True
```

### 3. Context Builder (`context_builder.py`)
Pass document context to expander:

```python
expansions = await expander.expand(
    query=user_query,
    context=self.get_collection_context()  # e.g., "Meeting notes from 2024"
)
```

## Configuration

```python
# SearchConfig
class SearchConfig:
    # Query expansion
    expansion_enabled: bool = True
    expansion_model: str = "gpt-4o-mini"
    expansion_max_variants: int = 6
    expansion_cache_ttl: int = 3600
    expansion_strong_signal_threshold: float = 0.85
```

## Files to Modify
- NEW: `src/nexus/search/query_expansion.py`
- `src/nexus/search/fusion.py` - Integrate expander
- `src/nexus/search/query_router.py` - Route expansion decisions
- `src/nexus/llm/context_builder.py` - Pass context to expander
- `src/nexus/search/__init__.py` - Export new module

## Acceptance Criteria
- [ ] `QueryExpander` class with LLM-based expansion
- [ ] Smart triggering (skip expansion on strong signal)
- [ ] Caching layer for expansion results
- [ ] Integration with existing hybrid search
- [ ] Configuration options (model, thresholds, cache TTL)
- [ ] Unit tests for expansion parsing
- [ ] Integration tests for search flow
- [ ] Benchmark: recall@10 improvement vs baseline

## Future Enhancements
- Fine-tuned local model (like QMD's 1.7B GRPO model)
- Reward-based training with rule-based scoring
- Entity preservation validation in expansions

## References
- QMD Query Expansion: https://github.com/tobi/qmd/tree/main/finetune
- HyDE Paper: https://arxiv.org/abs/2212.10496
- Related Issue: #1031 (Cross-encoder reranking)
